design-proposals/sriov-interfaces-mapping.md

```diff
@@ -0,0 +1,714 @@
+```yaml
+---
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: sriov-network-vlan100
+  namespace: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: kubevirt.io/sriov_net
+spec:
+  config: |-
+    {
+      "cniVersion":"0.3.1",
+      "name":"sriov-network-vlan100",
+      "type":"sriov",
+      "vlan":100,
+      "spoofchk":"on",
+      "trust":"off",
+      "vlanQoS":0,
+      "link_state":"enable",
+      "ipam":{}
+    }
+---
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: sriov-network-vlan200
+  namespace: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: kubevirt.io/sriov_net
+spec:
+  config: |-
+    {
+      "cniVersion":"0.3.1",
+      "name":"sriov-network-vlan200",
+      "type":"sriov",
+      "vlan":200,
+      "spoofchk":"on",
+      "trust":"off",
+      "vlanQoS":0,
+      "link_state":"enable",
+      "ipam":{}
+  }
+```
+3. The following environment variables will be present in the pod:
+```yaml
+# kubectl exec virt-launcher-sriovvmi2-wcszx -- env
+...
+PCIDEVICE_KUBEVIRT_IO_SRIOV_NET=0000:04:02.4,0000:04:02.5
+...
+KUBEVIRT_RESOURCE_NAME_sriovnet-vlan100=kubevirt.io/sriov_net
+KUBEVIRT_RESOURCE_NAME_sriovnet-vlan200=kubevirt.io/sriov_net
+```
+
+4. sriovnet-vlan100 and sriovnet-vlan200 networks will be represented by a Hostdev in the Domain XML:
+```yaml
+# kubectl exec -it virt-launcher-sriovvmi2-wcszx -- virsh dumpxml 1
+  <hostdev mode='subsystem' type='pci' managed='no'>
+    <alias name='ua-sriov-sriovnet-vlan100'/>
+    <driver name='vfio'/>
+    <source>
+          <address domain='0x0000' bus='0x04' slot='0x02' function='0x4'/>
+    </source>
+    <address type='pci' domain='0x0000' bus='0x06' slot='0x00' function='0x0'/>
+  </hostdev>
+  <hostdev mode='subsystem' type='pci' managed='no'>
+    <alias name='ua-sriov-sriovnet-vlan200'/>
+    <driver name='vfio'/>
+    <source>
+          <address domain='0x0000' bus='0x04' slot='0x02' function='0x5'/>
+    </source>
+    <address type='pci' domain='0x0000' bus='0x07' slot='0x00' function='0x0'/>
+  </hostdev>
+```
+
+#### Summary
+| SR-IOV VFs device pool name | kubevirt.io/sriov_net      |
+|:----------------------------|:---------------------------|
+| VM networks names           | sriovnet100, sriovnet200   |
+| Allocated VFs PCI addresses | 0000:04:03.6, 0000:04:03.7 |
+
+When HostDevice object is created its PCI address is picked as follows:
+networkToResource["sriovnet100"]  -------------------------> "kubevirt.io/sriov_net"
+addressesByResource["kubevirt.io/sriov_net"] ---------> "0000:04:03.6"
+
+There is no guarantee that "0000:04:03.6" is the correct PCI address for the network "sriovnet100"  and that the VF is configured with VLAN 100.
+
+This example shows that the current environment variable PCIDEVICE (given by sriov-dp) is simply not enough to directly map between the VM’s networks and  the VFs PCI address.
```

Comment:
Could you explain why this does not need a fix in sriov-dp? Do other applications in such a situation not care about knowing which device is which, if they have different configuration?

Reply:
Device plugins are executed before the CNI plugin does, in our care the sriov-dp execute first and allocates the VFs later on the sriov-cni is executed and configures the VFs according to the network config (passed by the NetworkAttachmentDefinition).
(I guess this is why those components are independent and not a single entity.)
At the time the sriov-dp is executed it doesn't have enough information to distinguish between the VFs it allocates, it doesn't have the network config.
The NPWG device-info doc explains it well:
https://github.com/k8snetworkplumbingwg/device-info-spec#current-state--limitations

Applications that need to distinguish between such devices, should expose the networks-status annotation through the downward API. This topic is also mentioned NPWG device info doc check this out: 
https://github.com/k8snetworkplumbingwg/device-info-spec#workflow-summary (last bullet).

Reply:
so, if the status annotation from multus already contains the information on the virt-launcher pod, why would we need our own virt-metadata annotation? I probably miss something.

Reply:
The networks-status annotation includes more information regarding the pod secondary NICs such as: NetworkAttachmentDefinition object names, the cluster CNI plugin name, pod network subnet, and basically whatever a CNI reporters according to the device-info spec.
For example: 
Consider the case where a VM is connected to the same network as the cluster nodes, that NICs status would be reflected on the networks-status annotation.
If there is a malicious entity inside the pod it could have access to that info if the annotation is mounted as is.

The idea is to extract only the relevant SR-IOV interfaces network-to-pci information and add it to the virt-metadata annotation that will be later on mounted to the pod.

Reply:
if there is security relevant info there, then better not to mount it. :+1: I would recommend the existing info flow then through virt-handler.

Reply:
Just to clarify, we are not exposing the network-status data, but rather a digested subset of it (network-pci map).